sklearn/metrics/_plot/confusion_matrix.py

```diff
@@ -89,14 +90,14 @@ def plot(self, include_values=True, cmap='viridis',
         if include_values:
             self.text_ = np.empty_like(cm, dtype=object)
             if values_format is None:
-                values_format = '.2g'
+                values_format = lambda x: 'd' if (x == 0 or log10(x) < 7) else '.2g'
 
             # print text with appropriate color depending on background
             thresh = (cm.max() + cm.min()) / 2.0
             for i, j in product(range(n_classes), range(n_classes)):
                 color = cmap_max if cm[i, j] < thresh else cmap_min
                 self.text_[i, j] = ax.text(j, i,
-                                           format(cm[i, j], values_format),
+                                           format(cm[i, j], values_format(cm[i, j])),
```

Comment:
the thing is, `values_format(cm[i, j])` will raise an error if the `values_format` is a user-defined format string like '{:.4f}'

Maybe just move the `if values_format is None:` right above, and don't define a lambda: just use `cm[i, j]` instead of `x`